acvm/src/pwg/block.rs

```diff
@@ -0,0 +1,187 @@
+use std::collections::{BTreeMap, HashMap};
+
+use acir::{
+    circuit::opcodes::{BlockId, BlockOp},
+    native_types::Witness,
+    FieldElement,
+};
+
+use crate::{GateResolution, OpcodeNotSolvable, OpcodeResolutionError};
+
+use super::{
+    arithmetic::{ArithmeticSolver, GateStatus},
+    directives::insert_witness,
+    expression_to_const,
+};
+
+#[derive(Default)]
+pub struct Blocks {
+    blocks: HashMap<BlockId, BlockSolver>,
+}
+
+impl Blocks {
+    pub fn solve(
+        &mut self,
+        id: BlockId,
+        trace: &Vec<BlockOp>,
+        solved_witness: &mut BTreeMap<Witness, FieldElement>,
+    ) -> Result<GateResolution, OpcodeResolutionError> {
+        let solver = self.blocks.entry(id).or_default();
+        solver.solve(solved_witness, trace)
+    }
+}
+
+#[derive(Default)]
+struct BlockSolver {
+    block_value: HashMap<u32, FieldElement>,
+    solved_operations: usize,
+}
+
+impl BlockSolver {
+    fn insert_value(
+        &mut self,
+        index: u32,
+        value: FieldElement,
+    ) -> Result<(), OpcodeResolutionError> {
+        let entry = self.block_value.entry(index);
+        match entry {
+            std::collections::hash_map::Entry::Vacant(e) => {
+                e.insert(value);
+            }
+            std::collections::hash_map::Entry::Occupied(e) => {
+                if e.get() != &value {
+                    return Err(OpcodeResolutionError::UnsatisfiedConstrain);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn get_value(&self, index: u32) -> Option<FieldElement> {
+        self.block_value.get(&index).copied()
+    }
+
+    // Try to solve block operations from the trace
+    // As long as operations are resolved, we update/read from the block_value
+    // We stop when an operation cannot be resolved
+    pub fn solve(
+        &mut self,
+        initial_witness: &mut BTreeMap<Witness, FieldElement>,
+        trace: &Vec<BlockOp>,
+    ) -> Result<GateResolution, OpcodeResolutionError> {
```

Comment:
I think this function generally suffers from a tower of if-lets which hampers readability. If we signalled that we should skip this opcode through an `OpcodeResolutionError` rather than `GateResolution`, we could cut this down using `.ok_or_else()?`

Reply:
If there is an error while we solve a gate, then we return the error, but not being able to solve a gate is not an error.

Reply:
Attempting to solve a gate and finding it to be unsolvable should be an error imo. It's a recoverable error as we may be able to solve it later but it's an error all the same.

Reply:
There is an important distinction between what you call a recoverable error and a definitive error. I explicit this distinction by using an additional type.

Reply:
I disagree and this is the expected usage of result. For instance if you want to open a file, it may not exist which is recoverable (as you can create the file) or you may not have permissions to open it which is unrecoverable. In both cases this is communicated though an Err though.

https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#matching-on-different-errors

Reply:
This is subject to discussion, the doc also says the following:

> Most languages don’t distinguish between these two kinds of errors and handle both in the same way, using mechanisms such as exceptions. Rust doesn’t have exceptions. Instead, it has the type Result<T, E> for recoverable errors and the panic! macro that stops execution when the program encounters an unrecoverable error. 

So according to the doc we should treat unsolvable as a (recoverable) error, while unsatisfied constraints should panic. That way these two kind of errors are indeed well separated.

Reply:
I think a better question is; what are the errors that are caused by a bug in this code, and what are the errors that are caused by a bug in user input. I class unrecoverable errors under errors which are caused by the acvm code, ie bugs and recoverable errors which are caused by the user input.

Since acvm cannot be sure that the user passed in a satisfiable witness assignment, these are classed as recoverable errors, so we return an error.
Similarly, since acvm cannot be sure that the user passed in an acvm circuit which is "correct" we also see this as a recoverable error and do not panic.

If ACVM was to check the circuit beforehand or have an API that only allowed one to build correct circuits, then it would be okay to panic everywhere else, since we can then assume that the user used our API and then comment that "This method will panic, if the circuit is not correct, which can be the case if you did not use the provided API to build ACIR"

So I guess the TLDR is that these are all users errors and it is upto Noir/frontend to panic since only on the Noir level do we know that this is actually a bug in the compiler.

Reply:
I agree to treat user errors as recoverable errors, using Result<,> mechanism, and to panic on compiler errors.
What we disagree on is that I think not solving a gate is not an error but a normal situation. It is definitively NOT a user error and also NOT a compiler error. What should be an error is when we cannot solve a circuit, which could be because of a user error (bad input) or an unknown bug.

Thank you @jfecher for the code snippet, we agreed to move this error handling out-of this PR, but I take note of your comments for the next PR.

Reply:
> What we disagree on is that I think not solving a gate is not an error but a normal situation.

My view on this is that it's definitely an error on the scale of these lower level functions (it's been provided inputs which it is unable to process) but it isn't an error on the compiler level (which is why we catch them there).

I'm going to resolve this comment and we can address `GateResolution` vs errors later.